library/src/main/java/com/datatorrent/lib/io/fs/AbstractFileInputOperator.java

```diff
@@ -710,11 +711,23 @@ public void emitTuples()
   }
 
   /**
+   * Notifies that the directory is being scanned.<br>
+   * Override this method to custom handling. Will be called once
+   */
+  protected void visitDirectory(Path filePath)
+  {
+  }
+
+  /**
    * Scans the directory for new files.
    */
   protected void scanDirectory()
   {
     if (System.currentTimeMillis() - scanIntervalMillis >= lastScanMillis) {
+      if (!visitedDirectory) {
```

Comment:
Could address this by storing a set of visited directory URIs and then relying on the provided DirectoryScanner to handle a recursive traversal if necessary